grc/upload/__init__.py

```diff
@@ -72,6 +103,70 @@ def uploadInfoPage(section_url: str):
     )
 
 
+@upload.route('/upload/<section_url>/document-password', methods=['GET', 'POST'])
+@LoginRequired
+def documentPassword(section_url: str):
+    section = next(filter(lambda section: section.url == section_url, sections), None)
+    if section is None:
+        abort(404)
+
+    deleteform = DeleteForm()
+    passwordForm = PasswordForm()
+    application_data = DataStore.load_application_by_session_reference_number()
+    error = ''
+    file_index = -1
+
+    if passwordForm.validate_on_submit():
+        files = section.file_list(application_data.uploads_data)
+        file_to_check = next(filter(lambda file: file.aws_file_name == passwordForm.file.data, files), None)
+        if file_to_check is not None:
+            data = AwsS3Client().download_object(file_to_check.aws_file_name)
+            if data is not None:
+                doc = fitz.open(stream=data.getvalue(), filetype='pdf')
+                if doc.needs_pass:
+                    if doc.authenticate(passwordForm.password.data):
+
+                        # Generate a new PDF
+                        import io
+                        from xhtml2pdf import pisa
+
+                        merger = fitz.open()
+                        merger.insert_pdf(doc)
+                        pdf = io.BytesIO()
+                        merger.save(pdf, deflate=True)
+                        merger.close()
+                        pdf.seek(0)
+
+                        AwsS3Client().delete_object(file_to_check.aws_file_name)
+                        AwsS3Client().upload_fileobj(pdf, file_to_check.aws_file_name)
+
+                        file_to_check.password_required = False
+                        DataStore.save_application(application_data)
+                    else:
+                        error = 'The password is incorrect'
+                        file_index = passwordForm.file_index.data
+                        passwordForm.password.errors.append(error)
+
+                doc.close()
+    elif passwordForm.file_index.data:
+        error = 'Enter your password'
+        file_index = passwordForm.file_index.data
+
+    files = [file for file in section.file_list(application_data.uploads_data) if file.password_required]
+    if len(files) == 0:
+        return local_redirect(url_for('upload.uploadInfoPage', section_url=section.url) + '#file-upload-section')
+
+    return render_template(
+        "upload/document-password.html",
+        deleteform=deleteform,
+        passwordForm=passwordForm,
+        section_url=section.url,
+        currently_uploaded_files=files,
+        error=error,
+        file_index=file_index
+    )
```

Comment:
Passing the error and file_index feels a bit clumsy.
I think we should consider re-structuring the workflow to make it fit more naturally into flask's framework